File: VidoX_Apple/VidoXProject/Services/Download/VideoRedownloader.swift
import Foundation
import SwiftData

/// Replaces the on-disk file for an existing library entry without creating a duplicate record.
enum VideoRedownloader {
    enum RedownloadError: LocalizedError {
        case invalidSourceURL
        case downloadNotAvailable
        case noFormat

        var errorDescription: String? {
            switch self {
            case .invalidSourceURL:
                return "This library item doesn’t have a valid source URL."
            case .downloadNotAvailable:
                return "This source can’t be downloaded automatically right now."
            case .noFormat:
                return "No downloadable format was found for this video."
            }
        }
    }

    /// Downloads the best available format for `video.sourceURL` and updates the same SwiftData row.
    @MainActor
    static func redownload(
        video: DownloadedVideo,
        modelContext: ModelContext,
        onProgress: @escaping (String, Double?) -> Void
    ) async throws {
        guard let sourceURL = URLNormalizer.url(from: video.sourceURL) else {
            throw RedownloadError.invalidSourceURL
        }

        onProgress("Looking up…", nil)
        let metadata = try await ExtractionRouter().extract(from: sourceURL)
        guard metadata.allowsRealDownload else {
            throw RedownloadError.downloadNotAvailable
        }
        guard let format = metadata.bestVideoFormat ?? metadata.formats.first(where: { !$0.isAudioOnly }) else {
            throw RedownloadError.noFormat
        }

        onProgress(metadata.usesYTDLP ? "Preparing…" : "Starting…", nil)

        var destination = FileStorage.makeVideoFileURL(
            preferredName: metadata.title.isEmpty ? video.title : metadata.title,
            fileExtension: format.fileExtension
        )

        let urlDownloader = URLSessionVideoDownloader()

        do {
            #if os(macOS)
            if metadata.platform == .youtube || (metadata.usesYTDLP && format.ytdlpFormatSelector != nil) {
                let ytdlpDownloader = YTDLPVideoDownloader()
                let selector = format.ytdlpFormatSelector
                    ?? ExperimentalSocialExtractor.youtubeFormatSelector
                for try await progress in ytdlpDownloader.download(
                    pageURL: metadata.sourceURL,
                    formatSelector: selector,
                    to: destination
                ) {
                    onProgress(
                        progress.totalBytes == nil ? "Downloading…" : "Finished",
                        progress.fraction
                    )
                }
                let directory = destination.deletingLastPathComponent()
                let stem = destination.deletingPathExtension().lastPathComponent
                if let match = try? FileManager.default.contentsOfDirectory(
                    at: directory,
                    includingPropertiesForKeys: nil
                )
                    .filter({ $0.lastPathComponent.hasPrefix(stem) })
                    .sorted(by: { $0.lastPathComponent.count < $1.lastPathComponent.count })
                    .first {
                    destination = match
                }
            } else if format.isHLSStream {
                onProgress("Exporting stream…", nil)
                let exportDestination = destination.deletingPathExtension().appendingPathExtension("mp4")
                try await HLSVideoDownloader.download(from: format.url, to: exportDestination)
                destination = exportDestination
                onProgress("Finished", 1)
            } else {
                try await downloadURLSession(
                    urlDownloader: urlDownloader,
                    from: format.url,
                    to: destination,
                    onProgress: onProgress
                )
            }
            #else
            if format.isHLSStream {
                onProgress("Exporting stream…", nil)
                let exportDestination = destination.deletingPathExtension().appendingPathExtension("mp4")
                try await HLSVideoDownloader.download(from: format.url, to: exportDestination)
                destination = exportDestination
                onProgress("Finished", 1)
            } else {
                try await downloadURLSession(
                    urlDownloader: urlDownloader,
                    from: format.url,
                    to: destination,
                    onProgress: onProgress
                )
            }
            #endif

            let previousPath = video.localFilePath
            let size = FileStorage.fileSize(at: destination)
            let storedPath = FileStorage.storedPath(for: destination)

            if !metadata.title.isEmpty {
                video.title = metadata.title
            }
            if let author = metadata.author {
                video.author = author
            }
            video.platformRaw = metadata.platform.rawValue
            video.localFilePath = storedPath
            video.fileExtension = destination.pathExtension.isEmpty
                ? format.fileExtension
                : destination.pathExtension
            video.fileSize = size
            video.downloadedAt = .now

            try modelContext.save()
            LocalSyncService.shared.noteUpserted(videoID: video.id)

            if previousPath != storedPath {
                FileStorage.removeFile(at: previousPath)
            }
        } catch {
            FileStorage.removeFile(at: destination.path)
            throw error
        }
    }

    private static func downloadURLSession(
        urlDownloader: URLSessionVideoDownloader,
        from remoteURL: URL,
        to destination: URL,
        onProgress: @escaping (String, Double?) -> Void
    ) async throws {
        for try await progress in urlDownloader.download(from: remoteURL, to: destination) {
            if let total = progress.totalBytes {
                let received = ByteCountFormatter.string(fromByteCount: progress.bytesReceived, countStyle: .file)
                let totalText = ByteCountFormatter.string(fromByteCount: total, countStyle: .file)
                onProgress("\(received) of \(totalText)", progress.fraction)
            } else {
                let received = ByteCountFormatter.string(fromByteCount: progress.bytesReceived, countStyle: .file)
                onProgress("\(received) downloaded", progress.fraction)
            }
        }
    }
}

File: VidoX_Google/app/src/main/java/com/ashbash/vidoxproject/services/download/VideoRedownloader.kt
package com.ashbash.vidoxproject.services.download

import android.app.Application
import com.ashbash.vidoxproject.VidoXApp
import com.ashbash.vidoxproject.data.DownloadedVideo
import com.ashbash.vidoxproject.models.VideoPlatform
import com.ashbash.vidoxproject.services.extraction.ExtractionError
import com.ashbash.vidoxproject.services.extraction.ExtractionRouter
import com.ashbash.vidoxproject.services.extraction.YOUTUBE_FORMAT_SELECTOR
import com.ashbash.vidoxproject.util.URLNormalizer
import com.ashbash.vidoxproject.util.VideoThumbnailLoader
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/** Replaces the on-disk file for an existing library entry without creating a duplicate. */
class VideoRedownloader(application: Application) {
    private val app = application as VidoXApp
    private val router = ExtractionRouter(application)
    private val urlDownloader = OkHttpVideoDownloader()
    private val ytdlpDownloader = YtDlpVideoDownloader(application)

    suspend fun redownload(
        video: DownloadedVideo,
        onProgress: (label: String, fraction: Float?) -> Unit
    ): DownloadedVideo = withContext(Dispatchers.IO) {
        suspend fun report(label: String, fraction: Float?) {
            withContext(Dispatchers.Main) { onProgress(label, fraction) }
        }

        val sourceURL = URLNormalizer.url(video.sourceURL)
            ?: throw ExtractionError.InvalidURL

        report("Looking up…", null)
        val metadata = router.extract(sourceURL)
        if (!metadata.allowsRealDownload) {
            throw IllegalStateException("This source can’t be downloaded automatically right now.")
        }
        val format = metadata.bestVideoFormat
            ?: metadata.formats.firstOrNull { !it.isAudioOnly }
            ?: throw IllegalStateException("No downloadable format was found for this video.")

        report(
            if (metadata.platform == VideoPlatform.YOUTUBE || metadata.usesYTDLP) "Preparing…" else "Starting…",
            null
        )

        var destination = app.fileStorage.makeVideoFile(
            if (metadata.title.isNotBlank()) metadata.title else video.title,
            format.fileExtension
        )

        try {
            if (metadata.platform == VideoPlatform.YOUTUBE ||
                (metadata.usesYTDLP && format.ytdlpFormatSelector != null)
            ) {
                val selector = format.ytdlpFormatSelector ?: YOUTUBE_FORMAT_SELECTOR
                ytdlpDownloader.download(
                    pageURL = metadata.sourceURL,
                    formatSelector = selector,
                    destination = destination
                ).collect { progress ->
                    report(
                        if (progress.totalBytes == null) "Downloading…" else "Finished",
                        progress.fraction
                    )
                }
                destination = ytdlpDownloader.resolveOutputFile(destination)
            } else {
                urlDownloader.download(format.url, destination).collect { progress ->
                    val label = if (progress.totalBytes != null) {
                        val received = app.fileStorage.formatByteCount(progress.bytesReceived)
                        val total = app.fileStorage.formatByteCount(progress.totalBytes)
                        "$received of $total"
                    } else {
                        "${app.fileStorage.formatByteCount(progress.bytesReceived)} downloaded"
                    }
                    report(label, progress.fraction)
                }
            }

            val thumbnailPath = runCatching {
                val remote = metadata.thumbnailURL ?: return@runCatching video.thumbnailPath
                val thumbFile = app.fileStorage.makeThumbnailFile()
                urlDownloader.download(remote, thumbFile).collect { }
                if (thumbFile.exists() && thumbFile.length() > 0L) {
                    video.thumbnailPath?.let { app.fileStorage.removeFile(it) }
                    app.fileStorage.storedThumbnailPath(thumbFile)
                } else {
                    thumbFile.delete()
                    video.thumbnailPath
                }
            }.getOrDefault(video.thumbnailPath)

            runCatching {
                VideoThumbnailLoader.cachedOrGenerate(
                    destination,
                    app.fileStorage.thumbnailsDirectory
                )
            }

            val size = app.fileStorage.fileSize(destination)
            val storedPath = app.fileStorage.storedPath(destination)
            val previousPath = video.localFilePath

            val updated = video.copy(
                title = metadata.title.ifBlank { video.title },
                author = metadata.author ?: video.author,
                platformRaw = metadata.platform.rawValue,
                thumbnailPath = thumbnailPath,
                localFilePath = storedPath,
                fileExtension = destination.extension.ifEmpty { format.fileExtension },
                downloadedAt = System.currentTimeMillis(),
                fileSize = size
            )
            app.repository.update(updated)

            if (previousPath != storedPath) {
                app.fileStorage.removeFile(previousPath)
            }
            updated
        } catch (e: Exception) {
            destination.delete()
            throw e
        }
    }
}
